source/common/json/json_streamer.h

```diff
@@ -0,0 +1,261 @@
+   *                 the entire json structure in memory before streaming it to
+   *                 the network.
+   */
+  explicit Streamer(Buffer::Instance& response) : response_(response) {}
+
+  class Array;
+  using ArrayPtr = std::unique_ptr<Array>;
+  class Map;
+  using MapPtr = std::unique_ptr<Map>;
+
+  /**
+   * Represents the current map or array. We keep track of what character is
+   * needed to close it, and whether or not the first entry has been added.
+   */
+  class Level {
+  public:
+    Level(Streamer& streamer, absl::string_view opener, absl::string_view closer);
+    virtual ~Level();
+
+    /**
+     * This must be called on the top level map or array. It's a programming
+     * error to call this method on a map that's not the top level.
+     * It's also a programming error to call this on map that isn't expecting
+     * a value. You must call Map::addKey prior to calling this.
+     *
+     * @return a newly created subordinate map, which becomes the new top level until destroyed.
+     */
+    MapPtr addMap();
+
+    /**
+     * This must be called on the top level map or array. It's a programming
+     * error to call this method on a map or array that's not the top level.
+     * It's also a programming error to call this on map that isn't expecting
+     * a value. You must call Map::addKey prior to calling this.
+     *
+     * @return a newly created subordinate array, which becomes the new top level until destroyed.
+     */
+    ArrayPtr addArray();
+
+    /**
+     * Adds a numeric value to the current array or map. It's a programming
+     * error to call this method on a map or array that's not the top level.
+     * It's also a programming error to call this on map that isn't expecting
+     * a value. You must call Map::addKey prior to calling this.
+     */
+    void addNumber(double d);
+
+    /**
+     * Adds a string constant value to the current array or map. The string
+     * will be sanitized per JSON rules.
+     *
+     * It's a programming error to call this method on a map or array that's not
+     * the top level. It's also a programming error to call this on map that
+     * isn't expecting a value. You must call Map::addKey prior to calling this.
+     */
+    void addString(absl::string_view str);
+
+  protected:
+    /**
+     * Initiates a new entry, serializing a comma separator if this is not the
+     * first one.
+     */
+    virtual void newEntry();
+
+    /**
+     * Renders a string or a double in json format. Doubles that are NaN are
+     * rendered as 'null'. Strings are json-sanitized if needed, and surrounded
+     * by quotes.
+     *
+     * @param Value the value to render.
+     * @return true if the value was a string, thus requiring a flush() before
+     *              returning control to the streamer client, which may mutate
+     *              the string.
+     */
+    void renderValue(const Value& value);
```

Comment:
Could you either name this addValue or name the others render*, for consistency?

While reviewing I code-searched for 'addValue' a couple of times expecting a parallel for 'addKey', and `addString` or `addNumber` are literally doing the 'render' operation for renderValue, so it's definitely not the only renderer.

My preference is for everything to be 'add', since addMap is only a partial render, but it *is* a partial render; naming only some of the things render would imply that nothing renders the map ends, and naming that one renderMap would be misleading because it doesn't render the whole map.

Reply:
changed newEntry and renderValue to addEntry and addValue. Good catch.

Reply:
Haha, oh no, I was actually more comfortable with newEntry, because that one *doesn't* add an entry, it at most adds a comma. With that one being add it reads weirdly because when you addEntries to a map, each entry now appears to add four things (addEntry, addSanitized, addEntry, addValue), even though an entry only has two things.

It's extra weird because Entry here isn't the singular of addEntries; we have "Entry" referring to "any key *or* value of a map or array", and "Entries" being "key-value pairs in a map, or values in an array".

Maybe rename `addEntry` -> `nextField`, or something like that?

Reply:
It is very useful to have a reviewer sanity-check naming that I had grown organically. Done!